### ISSUE TITLE
Stop upcasting result of torch.sum and torch.prod to int64

### ISSUE THREAD

**Issue by heitorschueroff** [pytorch/pytorch#62026]
The [Array API Standard](https://data-apis.org/array-api/latest/API_specification/statistical_functions.html#sum-x-axis-none-keepdims-false) specifies that the returned array from the operations `sum` and `prod` should have the same dtype as the input tensor.

**How is it currently done?**
Currently, both NumPy and PyTorch will promote signed integral types to `int64` to avoid overflow. Yet, for floating point types, they will return the same type as the input array instead of `float64`.

**How can it be fixed?**
Both NumPy and PyTorch already accept a `dtype` parameter which will cast the input and perform the reduction in the specified dtype. The intent of this is to pass a higher precision dtype in cases where overflow is a concern.

`sum` is the most popular reduction and one of the most popular operators. In order to deprecate this carefully, users would be required in a future release to specify the dtype parameter. This could mean changing their code in a lot of places which is quite disruptive.

**Comment by rgommers** [comment]
> Both NumPy and PyTorch already accept a `dtype` parameter which will cast the input and perform the reduction in the specified dtype. The intent of this is to pass a higher precision dtype in cases where overflow is a concern.

I don't think this is desired. It's still under discussion at https://github.com/data-apis/array-api/issues/202, and I believe @mruberry intends to deviate from the standard if it really would mandate `sum(int8) -> int8`.

**Comment by heitorschueroff** [comment]
@rgommers It makes sense for reductions that often result in a much larger value than the reduced values to return `int64`. For most use cases, this default behavior would be desirable. I'll update the title but keep this issue open to track when the array api standard has converged on an API for this.

**Comment by vadimkantorov** [comment]
Maybe related: https://github.com/pytorch/pytorch/issues/55366 in my view, ideally, torch.sum would allow a user to pass a target dtype, but would not materialize the upcasted input tensor

about defaults, i would suggest that PyTorch should set target dtype more practically, since torch.sum of bool tensors is often done to compute counts
